src/hotspot/share/gc/shared/partialArrayTaskStats.hpp

```diff
@@ -0,0 +1,114 @@
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ *
+ */
+
+#ifndef SHARE_GC_SHARED_PARTIALARRAYTASKSTATS_HPP
+#define SHARE_GC_SHARED_PARTIALARRAYTASKSTATS_HPP
+
+#include "logging/logHandle.hpp"
+#include "logging/logStream.hpp"
+#include "utilities/globalDefinitions.hpp"
+#include "utilities/growableArray.hpp"
+#include "utilities/macros.hpp"
+
+#if TASKQUEUE_STATS
+
+class outputStream;
+
+// Repository for collecting and reporting statistics about partial array task
+// processing.  Not thread-safe; each processing thread should have its own
+// stats object.
+class PartialArrayTaskStats {
+  size_t _split;
+  size_t _pushed;
+  size_t _stolen;
+  size_t _processed;
+
+  static LogTargetHandle log_target();
+  static bool is_log_enabled();
+  static void print_header(outputStream* s, const char* title);
+  void print_values(outputStream* s, uint id) const;
+  void print_total(outputStream* s) const;
+  void print_values_impl(outputStream* s) const;
+
+  void accumulate(const PartialArrayTaskStats& stats);
+
+public:
+  // All counters are initially zero.
+  PartialArrayTaskStats();
+
+  // Trivially copied and destroyed.
+
+  // Number of arrays split into partial array tasks.
+  size_t split() const { return _split; }
+
+  // Number of partial array tasks pushed onto a queue.
+  size_t pushed() const { return _pushed; }
+
+  // Number of partial array tasks stolen from some other queue.
+  size_t stolen() const { return _stolen; }
+
+  // Number of partial array tasks processed.
+  size_t processed() const { return _processed; }
+
+  void inc_split(size_t n = 1) { _split += n; }
+  void inc_pushed(size_t n = 1) { _pushed += n; }
+  void inc_stolen(size_t n = 1) { _stolen += n; }
+  void inc_processed(size_t n = 1) { _processed += n; }
+
+  // Set all counters to zero.
+  void reset();
+
+  // Log a table of statistics, if logging is enabled (gc+task+stats=trace).
+  //
+  // num_stats: The number of stats objects to include in the table, one row
+  // for each.
+  //
+  // access: A function taking a uint value < num_stats, and returning a
+  // pointer to the corresponding stats object.
+  //
+  // title: A string title for the table.
+  template<typename StatsAccess>
+  static void log_set(uint num_stats, StatsAccess access, const char* title);
+};
+
+template<typename StatsAccess>
+void PartialArrayTaskStats::log_set(uint num_stats,
```

Comment:
Can this be merged with its declaration? Seems kind of odd that these duplicates (method signature) are next to each other.

Reply:
That would implicitly declare it inline, which doesn't seem particularly
desirable here.  And it doesn't seem worth the overhead of splitting out into
a .inline.hpp file.  (That would let the logging includes be moved there,
rather than here in the .hpp file.  But that seems like a small benefit, since
I don't think there are going to be *that* many includes of this file.)

But the implicit inlining probably doesn't really matter after all, since the
access function is probably different in every use, so we'll so we'll have 1-1
uses to instantiations anyway.  So sure, merging.